### DISCUSSION TITLE
How to get old MCOE table with all generator attributes?

### DISCUSSION THREAD

**Discussion by bendnorman** [catalyst-cooperative/pudl#3336] [Help Me!] [answered]
In the `deployment-gap-model` project we've been using `PudlTabl.mcoe()` method:
https://github.com/deployment-gap-model-education-fund/deployment-gap-model/blob/a16f4dbfb09c90b4f1c5f72338a6514b904043a5/src/dbcp/etl.py#L77

This method returned a dataframe with 110 columns with all generator attributes. Which new PUDL tables should I use to grab all generator attributes? These are the columns in`mcoe_generators_yearly` and `mcoe_yearly`.

**I am using the database from the `v2023.12.01` release.** It looks like the `out_eia__yearly_generators` table from `nightly` contains all generator attributes.

## `mcoe_generators_yearly`
```
plant_id_eia
generator_id
report_date
unit_id_pudl
plant_id_pudl
plant_name_eia
utility_id_eia
utility_id_pudl
utility_name_eia
technology_description
energy_source_code_1
prime_mover_code
generator_operating_date
generator_retirement_date
operational_status
capacity_mw
fuel_type_code_pudl
planned_generator_retirement_date
capacity_factor
fuel_cost_from_eiaapi
fuel_cost_per_mmbtu
fuel_cost_per_mwh
heat_rate_mmbtu_mwh
net_generation_mwh
total_fuel_cost
total_mmbtu
```

## `mcoe_yearly`
```
plant_id_eia
generator_id
unit_id_pudl
report_date
capacity_factor
fuel_cost_from_eiaapi
fuel_cost_per_mmbtu
fuel_cost_per_mwh
heat_rate_mmbtu_mwh
net_generation_mwh
total_fuel_cost
total_mmbtu
```

**Comment by katie-lamb** [comment, marked as answer]
The new MCOE generators table in the latest version (`out_eia__yearly_generators`) does seem to truly have all the generator attributes. The previous version (`mcoe_generators_yearly`) only had the EIA generator attributes from `denorm_generators_eia` + the MCOE attributes that get merged on. The actual merge that's performed hasn't changed, only the input EIA generators attribute table, so it seems like the `denorm_generators_eia` table didn't have as many attributes as `_out_eia__yearly_generators` has. I'm not sure where those generator attributes would have lived in the older version if not in that denorm table, maybe the core table? What are some of the missing attributes and do they appear in other tables? I think to use that `mcoe_generators_yearly` table you'll need to do a PUDL `date_merge` with the tables that contain the missing attributes. Does that answer your question at all?

**Reply by bendnorman** [reply]
I think so! I remember deciding to add all generator attributes to `out_eia__yearly_generators` so the table is a one stop shop for generator information. It'd be nice to understand what actually changed in the code but for now I'll use `nightly` and wait for the new release.
